src/fields/field_balance.f90

```diff
@@ -0,0 +1,207 @@
+            ed = max(ed, dsqrt(tiny(ed)))
+            ep = max(ep, dsqrt(tiny(ep)))
+#endif
+            ! ef = 1 / (1 - exp[-2DK])
+            ef = one / (one - dexp(- two * kl * depth))
+
+#ifndef NDEBUG
+            ! To avoid "Floating-point exception - erroneous arithmetic operation"
+            ! when ef is really small.
+            ef = max(ef, dsqrt(tiny(ef)))
+#endif
+
+            !pq = exp[Kz] * (1 + exp[-2DK-2Kz]) / (KN * (1 - exp[-2DK]))
+            pq(:, ky, kx) = ef * ep * (one + ed) / (kl * bf)
+
+            !bq = exp[Kz] * (1 - exp[-2DK - 2Kz]) / (1 - exp[-2DK])
+            bq(:, ky, kx) = ef * ep * (one - ed)
+
+        end subroutine set_hyperbolic_functions
+
+        !::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::::
+
+        ! psi: streamfunction
+        ! Since the horizontal divergence u_x + v_y = 0, we can use a streamfunction to get u and v.
+        subroutine balance_fields(l_global)
+            logical, intent(in) :: l_global
+            double precision    :: psi(0:nz, box%lo(2):box%hi(2), box%lo(1):box%hi(1)) ! semi-spectral
+            double precision    :: ds(0:nz,  box%lo(2):box%hi(2), box%lo(1):box%hi(1)) ! semi-spectral
+            double precision    :: velbal(0:nz,  box%lo(2):box%hi(2), box%lo(1):box%hi(1), 3)
+            double precision    :: bbal(0:nz,  box%lo(2):box%hi(2), box%lo(1):box%hi(1))
+            integer             :: iz
+
+            !------------------------------------------------------------------
+            ! Define streamfunction and obtain balanced buoyancy anomaly:
+            call field_combine_semi_spectral(sbuoy)
+            do iz = 0, nz
+                psi(iz, :, :) = pq(iz, :, :) * sbuoy(nz, :, :)
+                ds(iz, :, :)  = bq(iz, :, :) * sbuoy(nz, :, :)
+            enddo
+            call field_decompose_semi_spectral(sbuoy)
+            call fftxys2p(ds, bbal)
+
+            !------------------------------------------------------------------
+            ! Obtain balanced velocity (u, v). Note: w = 0.
+            call diffy(psi, ds)
+            call fftxys2p(ds, velbal(:, :, :, 1))
+
+            ! u = - dpsi/dy
+            velbal(:, :, :, 1) = -velbal(:, :, :, 1)
+
+            ! v = dpsi/dx
+            call diffx(psi, ds)
+            call fftxys2p(ds, velbal(:, :, :, 2))
+
+            !------------------------------------------------------------------
+            ! Calculate balanced properties:
+
+            ! domain-averaged balanced kinetic energy: (note: for balanced fields w = 0)
+            kebal = get_horizontal_kinetic_energy(velbal, l_global)
+
+            ! domain-averaged balanced ape:
+            apebal = get_available_potential_energy(bbal, l_global, l_allreduce=.false.)
+
+            !------------------------------------------------------------------
+            ! Calculate imbalanced properties:
+
+            velbal(:, :, :, 3) = zero
+            velbal = velbal - vel
```

Comment:
Usually the imbalance = full - balanced; so vel - velbal here

Reply:
Yes, that is the wrong way round. I did it correct for the buoyancy. :)